operators/bosh-restore-operator/BoshRestoreService.js

```diff
@@ -0,0 +1,548 @@
+
+      const optionsData = _
+        .assign({
+          restoreMetadata: {
+            timeStamp: args.time_stamp,
+            filePath: _.get(service, 'restore_operation.filesystem_path'),
+            snapshotId: _.get(backupMetadata, 'snapshotId'),
+            deploymentName: deploymentName,
+            deploymentInstancesInfo: persistentDiskInfo,
+            baseBackupErrand: {
+              name: _.get(service, 'restore_operation.errands.base_backup_restore.name'),
+              instances: _.get(service, 'restore_operation.errands.base_backup_restore.instances')
+            },
+            pointInTimeErrand: {
+              name: _.get(service, 'restore_operation.errands.point_in_time.name'),
+              instances: _.get(service, 'restore_operation.errands.point_in_time.instances')
+            },
+            postStartErrand: {
+              name: _.get(service, 'restore_operation.errands.post_start.name'),
+              instances: _.get(service, 'restore_operation.errands.post_start.instances')
+            }
+          },
+          stateResults: {}
+        });
+
+      let restoreFileMetadata;
+      try {
+        restoreFileMetadata = await this.backupStore.getRestoreFile(data);
+      } catch(err) {
+        if (!(err instanceof errors.NotFound)) {
+          throw err;
+        }
+      }
+
+      await this.backupStore.putFile(_.assign(data, {
+        restore_dates: _.get(restoreFileMetadata, 'restore_dates')
+      }));
+
+      return eventmesh.apiServerClient.patchResource({
+        resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.RESTORE,
+        resourceType: CONST.APISERVER.RESOURCE_TYPES.DEFAULT_BOSH_RESTORE,
+        resourceId: opts.restore_guid,
+        options: optionsData,
+        status: {
+          'state': `${CONST.APISERVER.RESOURCE_STATE.IN_PROGRESS}_BOSH_STOP`,
+          'response': data
+        }
+      });
+    } catch (err) {
+      logger.error(`Error occurred while starting the restore: ${err}`);
+      return eventmesh.apiServerClient.updateResource({
+        resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.RESTORE,
+        resourceType: CONST.APISERVER.RESOURCE_TYPES.DEFAULT_BOSH_RESTORE,
+        resourceId: opts.restore_guid,
+        status: {
+          'state': CONST.APISERVER.RESOURCE_STATE.FAILED
+        }
+      });
+    }
+  }
+ 
+  async createPatchObject(opts, stateToUpdate) {
+    const options = _.assign({
+      service_id: opts.service_id,
+      plan_id: opts.plan_id,
+      tenant_id: opts.context ? this.getTenantGuid(opts.context) : opts.tenant_id,
+      instance_guid: opts.instance_guid
+    });
+    let restoreFileMetadata = await this.backupStore.getRestoreFile(options);
+    let restoreFinishiedAt = new Date().toISOString();
+    let restoreDates = _.get(restoreFileMetadata, 'restore_dates') || {};
+    let restoreDatesByState = _.get(restoreDates, stateToUpdate) || [];
+    if (_.indexOf(restoreDatesByState, restoreFinishiedAt) === -1) {
```

Comment:
Yep, this check is redundant. Had picked it from [here](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/operators/restore-operator/RestoreService.js#L82), where it made sense because the timestamp was being obtained from Agent (recorded once). But I think here we need to employ different strategy to ensure that this timestamp information does not get recorded more than once for single restore attempt, as this field is being used for restore quota validation.